File: tic-tac-toe.py
def print_board(board):
    """Prints the Tic-Tac-Toe board."""
    for row in board:
        print(" | ".join(row))
        print("-" * 5)

def check_winner(board, player):
    """Checks if the player has won."""
    # Check rows, columns, and diagonals
    for row in board:
        if all(cell == player for cell in row):
            return True
    for col in range(3):
        if all(row[col] == player for row in board):
            return True
    if all(board[i][i] == player for i in range(3)) or all(board[i][2-i] == player for i in range(3)):
        return True
    return False

def is_board_full(board):
    """Checks if the board is full."""
    return all(cell != ' ' for row in board for cell in row)

def get_move():
    """Gets the player's move and ensures it's valid."""
    while True:
        try:
            move = input("Enter your move (row and column: 1 2): ").split()
            row, col = int(move[0]) - 1, int(move[1]) - 1  # Convert to 0-based index
            if row not in range(3) or col not in range(3):
                print("Invalid move! Row and column must be between 1 and 3.")
            else:
                return row, col
        except (ValueError, IndexError):
            print("Invalid input! Please enter two numbers (row and column) separated by a space.")

def play_game():
    """Main function to play the Tic-Tac-Toe game."""
    board = [[' ' for _ in range(3)] for _ in range(3)]  # 3x3 empty board
    current_player = 'X'

    print("Welcome to Tic-Tac-Toe!")
    print_board(board)

    while True:
        print(f"Player {current_player}, it's your turn.")
        
        # Get the move from the player
        row, col = get_move()

        # Check if the chosen cell is already occupied
        if board[row][col] != ' ':
            print("This cell is already taken! Please choose another cell.")
            continue

        # Update the board with the current player's move
        board[row][col] = current_player
        print_board(board)

        # Check for a winner
        if check_winner(board, current_player):
            print(f"Player {current_player} wins!")
            break

        # Check for a draw
        if is_board_full(board):
            print("It's a draw!")
            break

        # Switch players
        current_player = 'O' if current_player == 'X' else 'X'

if __name__ == "__main__":
    play_game()

File: Tic-Tac-Toe.cpp
#include <iostream>
#include <vector>

using namespace std;

// Function to display the Tic-Tac-Toe board
void displayBoard(const vector<vector<char>>& board) {
    cout << "\n";
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            cout << board[i][j];
            if (j < 2) cout << " | ";
        }
        if (i < 2) cout << "\n---------\n";
    }
    cout << "\n";
}

// Function to check if there's a winner
char checkWinner(const vector<vector<char>>& board) {
    // Check rows and columns
    for (int i = 0; i < 3; ++i) {
        if (board[i][0] == board[i][1] && board[i][1] == board[i][2] && board[i][0] != ' ')
            return board[i][0];
        if (board[0][i] == board[1][i] && board[1][i] == board[2][i] && board[0][i] != ' ')
            return board[0][i];
    }

    // Check diagonals
    if (board[0][0] == board[1][1] && board[1][1] == board[2][2] && board[0][0] != ' ')
        return board[0][0];
    if (board[0][2] == board[1][1] && board[1][1] == board[2][0] && board[0][2] != ' ')
        return board[0][2];

    // No winner
    return ' ';
}

// Function to check if the board is full
bool isBoardFull(const vector<vector<char>>& board) {
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            if (board[i][j] == ' ')
                return false;
        }
    }
    return true;
}

// Main function for the game
int main() {
    vector<vector<char>> board(3, vector<char>(3, ' ')); // 3x3 board initialized with spaces
    char currentPlayer = 'X';
    char winner = ' ';
    int row, col;

    cout << "Welcome to Tic-Tac-Toe!\n";
    displayBoard(board);

    while (true) {
        cout << "Player " << currentPlayer << ", enter your move (row and column: 1 2 for row 1, col 2): ";
        cin >> row >> col;

        // Adjust for zero-based indexing
        row--;
        col--;

        // Check for valid move
        if (row < 0 || row >= 3 || col < 0 || col >= 3 || board[row][col] != ' ') {
            cout << "Invalid move. Try again.\n";
            continue;
        }

        // Make the move
        board[row][col] = currentPlayer;
        displayBoard(board);

        // Check for a winner
        winner = checkWinner(board);
        if (winner != ' ') {
            cout << "Player " << winner << " wins!\n";
            break;
        }

        // Check for a draw
        if (isBoardFull(board)) {
            cout << "It's a draw!\n";
            break;
        }

        // Switch players
        currentPlayer = (currentPlayer == 'X') ? 'O' : 'X';
    }

    cout << "Game over. Thanks for playing!\n";
    return 0;
}
